Mldbfb 601 http client integration

wsourdeau opened the pull request:
This PR provides an implementation of HTTPClient from upstream, as a milestone towards MLDBFB-239. The curl-based implementation was chosen as its performance was sufficiently good for S3 and as it is considered to be the most stable one.
Additionally:
* some related classes have been fixed after bugs were revealed
* libc_value_description has been added

jeremybarnes commented on http/http_client.h:
```diff
@@ -0,0 +1,432 @@
+#include <memory>
+#include <string>
+
+#include "mldb/ext/jsoncpp/value.h"
```
Would be good to not include the async_event_source header here, and to put more of the methods in the .cc file.

jeremybarnes commented on http/http_client.h:
```diff
@@ -0,0 +1,432 @@
+/* HTTP CLIENT IMPL                                                         */
+/****************************************************************************/
+
+struct HttpClientImpl : public AsyncEventSource {
```
Could we use composition rather than inheritance here?

jeremybarnes commented on http/http_client.h:
```diff
@@ -0,0 +1,432 @@
+/* HTTP CLIENT                                                              */
+/****************************************************************************/
+
+struct HttpClient : public AsyncEventSource {
```
Same here.

wsourdeau replied:
HttpClient is meant to be added to a MessageLoop (pending the asio version). In this case, it's required by the interface of MessageLoop::addSource.

jeremybarnes replied:
You could still decouple them, by having a getSource() method that returned a reference to an internal class that implemented MessageLoop::addSource.  The users of HttpClient shouldn't need to drag in files based upon the implementation.

wsourdeau replied:
Good idea. Then I could require the constructor to take a MessageLoop as parameter, similarly to how the EventLoop class is used. Would take make sense?

jeremybarnes replied:
Exactly.  And that parameter can be an opaque type.

jeremybarnes commented on http/http_client.h:
```diff
@@ -0,0 +1,432 @@
+       "implVersion": use version X of the HttpClientImpl, fallback
+       to HTTP_CLIENT_IMPL
+    */
+    HttpClient(const std::string & baseUrl,
```
Should all be in the .cc file.

wsourdeau replied:
Done. I have segregated most classes in their own files and put all the code in the .cc. Should make things much clearer and reduce the number of required includes.

jeremybarnes commented on http/http_client.h:
```diff
@@ -0,0 +1,432 @@
+/* This class is a child of HttpClientCallbacks and offers a simplified
+ * interface when support for progressive responses is not necessary. */
+
+struct HttpClientSimpleCallbacks : public HttpClientCallbacks
```
Why do we need a second class here?

wsourdeau replied:
HttpClientCallbacks is the full version that enables on-the-fly handling of requests and can therefore provide a version that is more memory efficient. The HttpClientSimpleCallbacks version enables easy handling of the requests through the need of a single callback with, in counterpart, the requirement to keep the full response body in memory.

jeremybarnes replied:
Why not make the HtpClientCallbacks implement the simple version by default?  Then we only have one class, and can override the other.

wsourdeau replied:
I could be it does mix both point of views in the same class, is that desirable?

jeremybarnes commented on http/tcp_socket_handler_impl.h:
```diff
@@ -64,6 +64,7 @@ struct TcpSocketHandlerImpl {
     typedef std::function<void(const boost::system::error_code & ec,
                                size_t bufferSize)> OnReadSome;
     OnReadSome onReadSome_;
+    bool closed_;
```
atomic?

wsourdeau replied:
Indeed!

jeremybarnes commented:
General comments
- Attempt to have as few includes and inlined methods as possible.  MLDB's compilation is slowed down by them and they tend to inroduce ABI couplings
- Please use Doxygen style /** comments for the method documentation, as Doxygen is used by the team to help generate code
- Please add a PR description (copied from the ticket is fine) as well as an indication as to what this will improve, and what the path forward is for further changesets that will come on top of this one

There were a few other minor comments in the code, but generally it seems to be in decent shape.

wsourdeau commented:
@jeremybarnes I have fixed the documentation for HttpClient, HttpClientImpl and LegacyEventLoop. Also, I have added a new module named "libc_value_description" with descriptions for struct provided by the libc (rusage, timeval).

If that looks fine to you, can I have your +1 ?

FinchPowers commented on http/http_client_impl_v1.cc:
```diff
@@ -0,0 +1,669 @@
+            afterContinue_ = false;
+    }
+    else {
+        const auto & cbs = request_->callbacks();
```
Je suppose que cette fonction aussi est noexcept?

FinchPowers commented on http/http_client_impl_v1.cc:
```diff
@@ -0,0 +1,669 @@
+            size_t oldTokenIdx(0);
+            size_t tokenIdx = headerLine.find(" ");
+            if (tokenIdx == string::npos || tokenIdx >= lineSize) {
+                throw ML::Exception("malformed header");
```
LOL. noexcept mais ça throw?!?

FinchPowers commented on http/http_client_impl_v1.cc:
```diff
@@ -0,0 +1,669 @@
+onCurlWrite(const char * data, size_t size)
+    noexcept
+{
+    request_->callbacks()->onData(*request_, data, size);
```
noexcept aussi?

FinchPowers commented on http/http_client_impl_v1.h:
```diff
@@ -0,0 +1,146 @@
+
+    /* AsyncEventSource */
+    virtual int selectFd() const;
+    virtual bool processOne();
```
override specifier si je ne me trompe pas?
http://en.cppreference.com/w/cpp/language/override

FinchPowers commented:
L'entête des fichiers devrait avoir le format suivant.
```
<filename>
<author>, <date>
This file is part of MLDB. Copyright <year> Datacratic. All rights reserved.
```

FinchPowers commented:
Donc, 
* quelques noexcept à vérifier
* des "override" à ajouter là où applicable
* les entêtes à uniformiser

Beau travail sur les tests. Je suppose que la couverture est bonne?

wsourdeau commented:
@FinchPowers pour les tests oui, la couverture est pratiquement complète. Chaque bug découvert a apporté son test depuis la version initiale, donc c'est relativement solide.

FinchPowers commented:
+1
